Add Option for Additional Ignored Folder

cabutlermit opened the pull request:
### Why these changes are being introduced

We need the ability to allow the web developer to optionally add an additional folder to the "exclude" list in the aws cli sync command. For example, this will allow for a documentation folder to be in the static site repo but not allow the contents of that folder to be synced to the website.

The changes here have been tested using the [future-of-libraries-static](https://github.com/MITLibraries/future-of-libraries-static) repo. The "dev" publishing workflow can be invoked manually, and it was tested twice, once with a value of `".docs/*"` passed to the shared workflow `EXCLUDE` input and once with nothing passed to the `EXCLUDE` input. The output from those two runs are available at the links below:

* [No value passed to the `EXCLUDE` variable](https://github.com/MITLibraries/future-of-libraries-static/actions/runs/3832485625/jobs/6522845274)
* [The `".docs/*"` value passed to the `EXCLUDE` variable](https://github.com/MITLibraries/future-of-libraries-static/actions/runs/3832477305/jobs/6522826280)

In both cases, go in to the workflow log and expand the "Sync S3 Content" step and expand the first line of that step to see the `if/then/else` statement that is executed.

The [future-of-libraries-static](https://github.com/MITLibraries/future-of-libraries-static) repo will get a PR after this one is approved and merged.


### How this addresses that need

* Add an optional workflow input ("EXCLUDE")
* Update the workflow sync step to use the additional EXCLUDE value if it is passed to this shared workflow


### Side effects of this change

None.


### Relevant ticket(s)

* https://mitlibraries.atlassian.net/browse/LM-118
* https://mitlibraries.atlassian.net/browse/LM-93


[LM-118]: https://mitlibraries.atlassian.net/browse/LM-118?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

cabutlermit committed:
Add Option for Additional Ignored Folder

Why these changes are being introduced:
We need the ability to allow the web developer to optionally add an
additional folder to the "exclude" list in the aws cli sync command.
For example, this will allow for a documentation folder to be in the
static site repo but not allow the contents of that folder to be
synced to the website.

How this addresses that need:
* Add an optional workflow input ("EXCLUDE")
* Update the workflow sync step to use the additional EXCLUDE value if
it is passed to this shared workflow

Side effects of this change:
None.

Relevant ticket(s):
* https://mitlibraries.atlassian.net/browse/LM-118
* https://mitlibraries.atlassian.net/browse/LM-93

JPrevost commented on .github/workflows/cdn-shared-publish.yml:
```diff
@@ -58,7 +61,12 @@ jobs:
-        run: aws s3 sync . ${{ inputs.S3URI }} --delete --exclude ".github/*" --exclude ".git/*" --exclude ".gitignore" 
+        run: |
+          if [ '${{ inputs.EXCLUDE }}' != '' ]; then
+            aws s3 sync . ${{ inputs.S3URI }} --delete --exclude ".github/*" --exclude ".git/*" --exclude ".gitignore" --exclude "${{ inputs.EXCLUDE }}"
```
I'm curious if we can come up with a way to allow multiple values to be passed, or to allow a full override.

Instead of 
```
aws s3 sync . ${{ inputs.S3URI }} --delete --exclude ".github/*" --exclude ".git/*" --exclude ".gitignore" --exclude "${{ inputs.EXCLUDE }}""
```

 maybe something like

```
aws s3 sync . ${{ inputs.S3URI }} --delete --exclude ".github/*" --exclude ".git/*" --exclude ".gitignore" "${{ inputs.ADDITIONAL_S3_SYNC_PARAMS }}"
```

and then you could pass in things like `--exclude docs --exclude another_folder --exclude yet_another_folder`.

I'm not sure it's the best implementation of that idea, but supporting "one more" folder seems like we'll run into edge cases that need "just one more" soon.

I'm okay with proceeding for now with only single folder override for now and returning to this if it ever comes up though fwiw.

cabutlermit replied:
I was sorta expecting this response from you. 😁 

I started with the quick solution because I didn't want to think about how to expand a list of folders/files to skip into multiple `--exclude` cli options. I hadn't actually thought of your approach -- don't think of it as an "exclude" but just make the web dev write out the desired list of cli options they want to tack on the end.

This is an easy enough change to make. I will push an update and let you know when it's ready for review. Given my afternoon schedule (I'm out-of-office for the afternoon) it'll likely be first thing tomorrow morning.

cabutlermit replied:
@JPrevost I made the change (see the WIP commit 42d07d2). I did a test run from the future-of-libraries-static repo (see the output [here](https://github.com/MITLibraries/future-of-libraries-static/actions/runs/3841012558/jobs/6540742065)) and it worked as expected.

The old `EXCLUDE` line was replaced with `SYNC_PARAMS: --exclude ".docs/*" --exclude "README.md"`.

JPrevost replied:
Looks exactly like what I was looking for. Thanks!

cabutlermit committed:
Switch to Addtional Sync Parameters

Why these changes are being introduced:
The original attempt was restricted to adding a single excluded file
or folder. This was too limiting. This change addresses a concern
raised by the reviewer that we will likely need more flexibility than
just one file or folder.

How this addresses that need:
* Instead of a single `--exclude`, the input variable is configured as
a string to tack on the end of the `aws s3 sync` command.
* Update README for new input values for this workflow

Side effects:
Any pre-existing caller workflows will need to be updated to match
the new input value name.

Relevant tickets:
* https://mitlibraries.atlassian.net/browse/LM-93

cabutlermit commented:
@JPrevost I cleaned up my last commit message and updated the README and did a little commit squashing. For consistency, I'm going to re-request your review on this (so it doesn't look like I made changes after you approved this). 😁
